### PR TITLE
[14.0][FIX] base: Prevent singleton ValueError on res.partner's barcode

### LINKED ISSUE #137141: [16.0] _check_barcode_unicity on res.partner doesn't handle multiple records
Impacted versions: 16.0

Steps to reproduce:

```python
self.env["res.partner"].search([]).write({"barcode": "foo"})
```

Current behavior:

ValueError: Expected singleton

Expected behavior:

No error

Video/Screenshot link (optional):

Support ticket number submitted via odoo.com/help (optional):

### PR DESCRIPTION
**Pull request by carmenbianca**
Description of the issue/feature this PR addresses:

Fixes #137141 

Current behavior before PR:

Singleton ValueError

Desired behavior after PR is merged:

No error, or a ValidationError

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ryv-odoo** [comment]
Hello @carmenbianca ,
Nice catch, it would be nice to target the 14.0, pretty sure that it doesn't work either (the issue coming from https://github.com/odoo/odoo/pull/41658).

Thank you.

**Comment by carmenbianca** [comment]
Hi @ryv-odoo 

Is there anything I ought do to also target 14.0 and 15.0? I'm not familiar with the magic Odoo uses to backport/forwardport PRs.

**Comment by carmenbianca** [comment]
@xmo-odoo I targeted `14.0` with this PR. Ought I create new PRs for `15.0` and `16.0`?

**Comment by carmenbianca** [comment]
@ryv-odoo Not sure what the conflicting files are about. It's a single commit touching a single file on top of the HEAD of 14.0.

File: odoo/addons/base/models/res_partner.py [hunk 1]
```diff
@@ -425,8 +425,9 @@ def onchange_company_type(self):
 
     @api.constrains('barcode')
     def _check_barcode_unicity(self):
-        if self.barcode and self.env['res.partner'].search_count([('barcode', '=', self.barcode)]) > 1:
-            raise ValidationError('An other user already has this barcode')
+        for partner in self:
+            if partner.barcode and self.env['res.partner'].search_count([('barcode', '=', partner.barcode)]) > 1:
```
Comment (HydrionBurst) [inline odoo/addons/base/models/res_partner.py:429]: @ryv-odoo
`search_count` in a loop is not ideal, but may not be the performance bottleneck in the business
But I am curious in https://github.com/odoo/odoo/pull/41658, if the unique constraint is also used for search before the PR.
maybe we can index the company dependent field in ir.property with the help of the our new index ?@xmo-odoo

Comment (carmenbianca) [reply odoo/addons/base/models/res_partner.py:429]: I wouldn't do premature optimisation here. In 99,9% of cases, `self` will be a list of only 1 record. This PR fixes an extreme edge case.

Comment (HydrionBurst) [reply odoo/addons/base/models/res_partner.py:429]: > irproperty already indexes all relevant fields

I mean when you search a `partner.search(['barcode', '=', 'foo'])`
The sql would be something like
```
SELECT "ir_property"."id" FROM "ir_property" WHERE ((("ir_property"."fields_id" = 890) AND (("ir_property"."company_id" in (1)) OR "ir_property"."company_id" IS NULL)) AND ("ir_property"."value_text" = 'foo')) ORDER BY  "ir_property"."id"
```
`name` and `res_id` are indexed in the `ir.property` but not the `value_text`

I just check with the pos team, at least for know they don't receive any performance complain about the barcode search for business. So it is OK for now.

Comment (ryv-odoo) [reply odoo/addons/base/models/res_partner.py:429]: Indeed @HydrionBurst , but this code has four years and nobody complains about it. I am not a huge fan to change more than necessary in stable, it is why I didn't ask to change that.

**Pull request closed without merging**